components/common-go/baseserver/server.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package baseserver
+
+import (
+	"context"
+	"fmt"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"net"
+	"net/http"
+	"os"
+	"os/signal"
+	"syscall"
+)
+
+func New(name string, opts ...Option) (*Server, error) {
+	cfg, err := evaluateOptions(defaultConfig(), opts...)
+	if err != nil {
+		return nil, fmt.Errorf("invalid config: %w", err)
+	}
+
+	server := &Server{
+		Name: name,
+		cfg:  cfg,
+	}
+
+	if initErr := server.initializeHTTP(); initErr != nil {
+		return nil, fmt.Errorf("failed to initialize http server: %w", initErr)
+	}
+	if initErr := server.initializeGRPC(); initErr != nil {
+		return nil, fmt.Errorf("failed to initialize grpc server: %w", initErr)
+	}
+
+	return server, nil
+}
+
+type Server struct {
+	// Name is the name of this server, used for logging context
+	Name string
+
+	cfg *config
+
+	// http is an http Server
+	http         *http.Server
+	httpMux      *http.ServeMux
+	httpListener net.Listener
+
+	// grpc is a grpc Server
+	grpc         *grpc.Server
+	grpcListener net.Listener
+
+	// listening indicates the server is serving. When closed, the server is in the process of graceful termination.
+	listening chan struct{}
+}
+
+func (s *Server) ListenAndServe() error {
+	var err error
+	s.grpcListener, err = net.Listen("tcp", fmt.Sprintf(":%d", s.cfg.grpcPort))
+	if err != nil {
+		return fmt.Errorf("failed to acquire port %d", s.cfg.grpcPort)
+	}
+
+	s.httpListener, err = net.Listen("tcp", fmt.Sprintf(":%d", s.cfg.httpPort))
+	if err != nil {
+		return fmt.Errorf("failed to acquire port %d", s.cfg.grpcPort)
+	}
+
+	errors := make(chan error)
+	defer close(errors)
+	s.listening = make(chan struct{})
+
+	go func() {
+		s.Logger().
+			WithField("protocol", "grpc").
+			Infof("Serving gRPC on %s", s.grpcListener.Addr().String())
+		if serveErr := s.grpc.Serve(s.grpcListener); serveErr != nil {
+			if s.isClosing() {
+				return
+			}
+
+			errors <- serveErr
+		}
+	}()
+
+	go func() {
+		s.Logger().
+			WithField("protocol", "http").
+			Infof("Serving http on %s", s.httpListener.Addr().String())
+		if serveErr := s.http.Serve(s.httpListener); serveErr != nil {
+			if s.isClosing() {
+				return
+			}
+
+			errors <- serveErr
+		}
+	}()
+
+	signals := make(chan os.Signal, 1)
```

Comment:
I don't think you need to specify a channel buffer size here.

Reply:
I was going by this example: https://gobyexample.com/signals

I think it doesn't hurt to specify it as that we're ensuring that repeated signals (unsure if we'd receive them) don't get buffered and block instead. That's desirable as we're looking to limit memory usage and complete our termination procedure. In practice, in k8s, we'll receive SIGINT, and if we don't shut down in time k8s will just terminate the pod entirely